core/src/main/java/org/elasticsearch/ingest/processor/DateFormat.java

```
@@ -0,0 +1,99 @@
+
+import java.util.Locale;
+import java.util.Optional;
+import java.util.function.Function;
+
+enum DateFormat {
+    Iso8601 {
+        @Override
+        Function<String, DateTime> getFunction(DateTimeZone timezone) {
+            return ISODateTimeFormat.dateTimeParser().withZone(timezone)::parseDateTime;
+        }
+    },
+    Unix {
+        @Override
+        Function<String, DateTime> getFunction(DateTimeZone timezone) {
+            return (date) -> new DateTime((long)(Float.parseFloat(date) * 1000), timezone);
+        }
+    },
+    UnixMs {
+        @Override
+        Function<String, DateTime> getFunction(DateTimeZone timezone) {
+            return (date) -> new DateTime(Long.parseLong(date), timezone);
+        }
+
+        @Override
+        public String toString() {
+            return "UNIX_MS";
+        }
+    },
+    Tai64n {
+        @Override
+        Function<String, DateTime> getFunction(DateTimeZone timezone) {
+            return (date) -> new DateTime(parseMillis(date), timezone);
+        }
+
+        private long parseMillis(String date) {
+            if (date.startsWith("@")) {
+                date = date.substring(1);
+            }
+            long base = Long.parseLong(date.substring(1, 16), 16);
+            // 1356138046000
+            long rest = Long.parseLong(date.substring(16, 24), 16);
+            return ((base * 1000) - 10000) + (rest/1000000);
+        }
+    };
+
+    abstract Function<String, DateTime> getFunction(DateTimeZone timezone);
+
+    static Optional<DateFormat> fromString(String format) {
+        switch (format) {
+            case "ISO8601":
+                return Optional.of(Iso8601);
+            case "UNIX":
+                return Optional.of(Unix);
+            case "UNIX_MS":
+                return Optional.of(UnixMs);
+            case "TAI64N":
+                return Optional.of(Tai64n);
+            default:
+                return Optional.empty();
```

should this rather throw an exception?